packages/go_router/lib/src/route.dart

```diff
@@ -458,4 +558,771 @@ class ShellRoute extends RouteBase {
+      ShellRouteContext shellRouteContext) {
+    if (pageBuilder != null) {
+      return pageBuilder!(
+          context, state, _createShell(context, shellRouteContext));
+    }
+    return null;
+  }
+
+  @override
+  GlobalKey<NavigatorState> navigatorKeyForSubRoute(RouteBase subRoute) {
+    final StatefulShellBranch? branch = branches.firstWhereOrNull(
+        (StatefulShellBranch e) => e.routes.contains(subRoute));
+    assert(branch != null);
+    return branch!.navigatorKey;
+  }
+
+  @override
+  Iterable<GlobalKey<NavigatorState>> get _navigatorKeys =>
+      branches.map((StatefulShellBranch b) => b.navigatorKey);
+
+  StatefulNavigationShell _createShell(
+          BuildContext context, ShellRouteContext shellRouteContext) =>
+      StatefulNavigationShell(
+          shellRouteContext: shellRouteContext,
+          router: GoRouter.of(context),
+          containerBuilder: navigatorContainerBuilder);
+
+  static Widget _indexedStackContainerBuilder(BuildContext context,
+      StatefulNavigationShell navigationShell, List<Widget> children) {
+    return _IndexedStackedRouteBranchContainer(
+        currentIndex: navigationShell.currentIndex, children: children);
+  }
+
+  static List<RouteBase> _routes(List<StatefulShellBranch> branches) =>
+      branches.expand((StatefulShellBranch e) => e.routes).toList();
+
+  static Set<GlobalKey<NavigatorState>> _debugUniqueNavigatorKeys(
+          List<StatefulShellBranch> branches) =>
+      Set<GlobalKey<NavigatorState>>.from(
+          branches.map((StatefulShellBranch e) => e.navigatorKey));
+
+  static bool _debugValidateParentNavigatorKeys(
+      List<StatefulShellBranch> branches) {
+    for (final StatefulShellBranch branch in branches) {
+      for (final RouteBase route in branch.routes) {
+        if (route is GoRoute) {
+          assert(route.parentNavigatorKey == null ||
+              route.parentNavigatorKey == branch.navigatorKey);
+        }
+      }
+    }
+    return true;
+  }
+
+  static bool _debugValidateRestorationScopeIds(
+      String? restorationScopeId, List<StatefulShellBranch> branches) {
+    if (branches
+        .map((StatefulShellBranch e) => e.restorationScopeId)
+        .whereNotNull()
+        .isNotEmpty) {
+      assert(
+          restorationScopeId != null,
+          'A restorationScopeId must be set for '
+          'the StatefulShellRoute when using restorationScopeIds on one or more '
+          'of the branches');
+    }
+    return true;
+  }
+}
+
+/// Representation of a separate branch in a stateful navigation tree, used to
+/// configure [StatefulShellRoute].
+///
+/// The only required argument when creating a StatefulShellBranch is the
+/// sub-routes ([routes]), however sometimes it may be convenient to also
+/// provide a [initialLocation]. The value of this parameter is used when
+/// loading the branch for the first time (for instance when switching branch
+/// using the goBranch method in [StatefulNavigationShell]).
```

Comment:
How do you add a custom transition animation for two StatefulShellBranches? I tried to use pageBuilder with a FadeTransitionPage and it didn't work for me.

Reply:
The sample code did include an example on how to accomplish this, but it was removed to reduce simplify the sample code. I guess this could be included in the `custom_stateful_shell_route.dart` sample though. 

But basically, switching branch (i.e. parallel Navigator) is different that simply navigating to a new route within a Navigator when it comes to animations (i.e. this is different than push/pop within a single Navigator - a good comparison is CupertinoTabBar/CupertinoTabScaffold). Therefore, it is the custom shell implementation that is responsible for animating branch switches. Perhaps this needs to be mentioned somewhere in the documentation.

Reply:
Docs would be great, an example would be great too if you have one.